nmostesting/NMOSTesting.py

```diff
@@ -340,6 +341,18 @@
     #     }],
     #     "class": IS1001Test.IS1001Test
     # },
+    "IS-12-01": {
+        "name": "IS-12 NMOS Control Protocol",
+        "specs": [{
+            "spec_key": "is-04",
+            "api_key": "node",
+        }, {
+            "spec_key": "is-12",
+            "api_key": "control",
+            "disable_fields": ["host", "port"]

```

Comment:
Normally we'd make/allow the test tool user enter the host/port for the control endpoint and then validate in the "interoperability with IS-04" test case that that is indeed the endpoint advertised for at least one Device in the Node API. Would that work for IS-12?

Reply:
We can certainly use the host and port to validate the control endpoint. However, I don't think we can "calculate" the control endpoint href simply from the host and port as the actual href can be arbitrary (so far according to the IS-12 spec e.g. `ws://127.0.0.1:8080/x-nmos/ncp/v1.0/connect`). AFAICS neither IS-05 or IS-07, for example, constrain the form of the href, or am I missing something there?

Reply:
IS-05 and IS-07 actually both *do* constrain the URL path.

However, IS-08 doesn't (fully), it introduces the concept of a "selector" which is appended to the defined base path.
You can see how that's represented in the `TEST_DEFINITIONS` for the IS-08-02 test suite with `"selector": True` (disabled for the IS-04 API).

However, that's not quite enough for you here, where there's no defined base path, so this codeis wrong:

https://github.com/AMWA-TV/nmos-testing/blob/5114d1f79d9dd4f359c0e41ad787ff8c3fca7c60/nmostesting/NMOSTesting.py#L559-L564

The IS-10-01 suite actually needs the same as IS-12, but that test suite is languishing since the IS-10 spec change to move from a completely fixed path to no defined base path in common with the underlying [OAuth 2.0 Authorization Server Metadata](https://www.rfc-editor.org/rfc/rfc8414) spec...

Changing `selector` from Boolean to an enum None/Append/Replace could work for both IS-12 and IS-10... but really `selector` ought to be a property of each API so we could handle test suites needing two APIs with different selector modes. But then we need to gather the UI requirements differently... :-/

Reply:
Note that the referenced code in NMOSTesting.py is constructing an http/https url to an API, whereas here we want to construct a ws/wss url - I could add a `"websocket": True` to the test definition (similar to the `"selector": True`) so the url can be constructed correctly.

A question for the IS-12 spec is whether the url *should* (must) be constrained or whether that is too restrictive - I would argue for consistency with IS-08 *at least* and so the existing selector mechanism would work:
```
api_key: "ncp",
version: "v1.0",
selector: True,
websocket: True
```
with parameters captured from the UI:
```
host: "127.0.0.1"
port: "8080"
API selector: "connect"
```
Would yield the desired url as specified in the example mock device (`ws://127.0.0.1:8080/x-nmos/ncp/v1.0/connect`)

Reply:
You're right about protocol for sure. Thinking about adding`websocket` at the same spec level as the current attributes rather than per api clearly demonstrates that it's the wrong level.

Reply:
Thinking on this, there is a difference between the HTTP endpoints and the WebSocket ones that justifies the former having fixed URL paths and the latter not needing to. The WebSocket endpoint supports a single request, opening a connection over which subsequent 'requests' are made. In the other hand the HTTP URL is a base from which paths to many endpoints are determined. The base path could have been allowed to be anything since the other endpoints could be (and may well be in current implementations)  calculated via relative paths, so it's not much of a distinction to be sure!

Reply:
Yes, I see the distinction, and it's a pair point.

I think there are two decisions to be made here:
1) Should the WebSocket URL have a known form that is consistent with the existing IS HTTP URLs (i.e. `ws(s)://<ip address or hostname>:<port>/x-nmos/<api type>/<api version>/[<api selector>/]`?  My feeling on this is yes for consistency and to make development and deployment easier. Whether an api selector is needed or desirable is an open question.
2) Should NMOSTesting "calculate" this URL in the same way it calculates the base URL of the HTTP endpoints. Here either we use the existing mechanism with the tweaks suggested in the [current PR](https://github.com/AMWA-TV/nmos-testing/pull/810), or we implement a separate mechanism that is very similar due to the distinction you've highlighted above. Again my feeling is to reuse the existing mechanism as per the current state of the PR.